lapack/testlapack/dorgr2.go

```diff
@@ -0,0 +1,83 @@
+// Copyright ©2021 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package testlapack
+
+import (
+	"fmt"
+	"testing"
+
+	"golang.org/x/exp/rand"
+	"gonum.org/v1/gonum/blas"
+	"gonum.org/v1/gonum/blas/blas64"
+	"gonum.org/v1/gonum/lapack"
+)
+
+type Dorgr2er interface {
+	Dorgr2(m, n, k int, a []float64, lda int, tau []float64, work []float64)
+
+	Dgerqfer
+}
+
+func Dorgr2Test(t *testing.T, impl Dorgr2er) {
+	rnd := rand.New(rand.NewSource(1))
+	for _, k := range []int{0, 1, 2, 5} {
+		for _, m := range []int{k + 1, k + 2, k + 4} {
+			for _, n := range []int{m, m + 1, m + 2, m + 4, m + 7} {
+				for _, lda := range []int{n, n + 5} {
+					dorgr2Test(t, impl, rnd, m, n, k, lda)
+				}
+			}
+		}
+	}
+
+}
+
+// dorgr2Test is a single case, random test of Dorgr2 implementation. m <= n.
+func dorgr2Test(t *testing.T, impl Dorgr2er, rnd *rand.Rand, m, n, k, lda int) {
+	const tol = 1e-12
+	name := fmt.Sprintf("n=%v,m=%v,lda=%v,k=%v", n, m, lda, k)
+	if lda == 0 {
+		lda = n
+	}
+	// k = min(n, m)
+	a := randomGeneral(m, n, lda, rnd)
+	aCopy := cloneGeneral(a)
+	// Compute the RQ decomposition of A.
+	tau := make([]float64, min(m, n))
+	work := make([]float64, 1)
+	impl.Dgerqf(m, n, a.Data, a.Stride, tau, work, -1)
+	work = make([]float64, int(work[0]))
+	impl.Dgerqf(m, n, a.Data, a.Stride, tau, work, len(work))
+
+	// Generate the upper triangular matrix R from subarray A[0:m, n-m:n]
+	r := zeros(m, m, m)
+	for i := 0; i < m; i++ {
+		for j := n - m; j < n; j++ {
+			jr := j - (n - m)
+			if i > jr {
+				continue // lower triangle is zero
+			}
+			r.Data[i*r.Stride+jr] = a.Data[i*a.Stride+j]
+		}
+	}
+	// Compute the matrix Q using Dorg2r.
+	impl.Dorgr2(m, n, k, a.Data, a.Stride, tau, work)
+	if m == 0 {
+		return
+	}
+	q := a
+	// Test Q orthogonality.
+	res := residualOrthogonal(q, true)
+	if res > tol {
+		t.Errorf("%v: |I - Q * Qᵀ| residual too large (%g)", name, res)
+	}
+
+	// Test |A - R*Q| is small
+	blas64.Gemm(blas.NoTrans, blas.NoTrans, 1, r, q, -1, aCopy)
```

Comment:
With `Dorgr2` you can only reconstruct the last k rows of A, so you must only test that
```
|A[m-k:m,0:n] - R[m-k:m,0:m] * Q|
```
is small.